apple-dmg/src/lib.rs

```diff
@@ -0,0 +1,346 @@
+            }
+            ChunkType::Raw => Ok(Box::new(compressed_chunk)),
+            ChunkType::Zlib => Ok(Box::new(ZlibDecoder::new(compressed_chunk))),
+            ChunkType::Adc | ChunkType::Bzlib | ChunkType::Lzfse => unimplemented!(),
+            ChunkType::Term => Ok(Box::new(std::io::empty())),
+        }
+    }
+
+    pub fn data_checksum(&mut self) -> Result<u32> {
+        self.r.seek(SeekFrom::Start(self.koly.data_fork_offset))?;
+        let mut data_fork = Vec::with_capacity(self.koly.data_fork_length as usize);
+        (&mut self.r)
+            .take(self.koly.data_fork_length)
+            .read_to_end(&mut data_fork)?;
+        Ok(crc32fast::hash(&data_fork))
+    }
+
+    pub fn partition_table(&self, i: usize) -> Result<BlkxTable> {
+        self.plist().partitions()[i].table()
+    }
+
+    pub fn partition_name(&self, i: usize) -> &str {
+        &self.plist().partitions()[i].name
+    }
+
+    pub fn partition_data(&mut self, i: usize) -> Result<Vec<u8>> {
+        let table = self.plist().partitions()[i].table()?;
+        let mut partition = vec![];
+        for chunk in &table.chunks {
+            std::io::copy(&mut self.sector(chunk)?, &mut partition)?;
+        }
+        Ok(partition)
+    }
+}
+
+pub struct DmgWriter<W: Write + Seek> {
+    xml: Plist,
+    w: W,
+    data_hasher: Hasher,
+    main_hasher: Hasher,
+    sector_number: u64,
+    compressed_offset: u64,
+}
+
+impl DmgWriter<BufWriter<File>> {
+    pub fn create(path: &Path) -> Result<Self> {
+        let w = BufWriter::new(File::create(path)?);
+        Ok(Self::new(w))
+    }
+}
+
+impl<W: Write + Seek> DmgWriter<W> {
+    pub fn new(w: W) -> Self {
+        Self {
+            xml: Default::default(),
+            w,
+            data_hasher: Hasher::new(),
+            main_hasher: Hasher::new(),
+            sector_number: 0,
+            compressed_offset: 0,
+        }
+    }
+
+    pub fn create_fat32(mut self, fat32: &[u8]) -> Result<()> {
+        anyhow::ensure!(fat32.len() % 512 == 0);
+        let sector_count = fat32.len() as u64 / 512;
+        let mut mbr = ProtectiveMBR::new();
+        let mut partition = PartRecord::new_protective(Some(sector_count.try_into()?));
+        partition.os_type = 11;
+        mbr.set_partition(0, partition);
+        let mbr = mbr.as_bytes()?;
+        self.add_partition("Master Boot Record (MBR : 0)", &mbr)?;
+        self.add_partition("FAT32 (FAT32 : 1)", fat32)?;
+        self.finish()?;
+        Ok(())
+    }
+
+    pub fn add_partition(&mut self, name: &str, bytes: &[u8]) -> Result<()> {
+        anyhow::ensure!(bytes.len() % 512 == 0);
+        let id = self.xml.partitions().len() as u32;
+        let name = name.to_string();
+        let mut table = BlkxTable::new(id, self.sector_number, crc32fast::hash(bytes));
+        for chunk in bytes.chunks(2048 * 512) {
+            let mut encoder = ZlibEncoder::new(chunk, Compression::best());
```

Comment:
The compression format probably needs to be configurable. But we can enhance this post merge.